app/decorators/spree/line_item_decorator.rb

```diff
@@ -4,6 +4,21 @@ def self.prepended(base)
       base.has_many :ad_hoc_option_values_line_items, dependent: :destroy
       base.has_many :ad_hoc_option_values, through: :ad_hoc_option_values_line_items
       base.has_many :product_customizations, dependent: :destroy
+
+      base.validate :product_customization_should_exist_if_required
+    end
+
+    def product_customization_should_exist_if_required
+      existing_product_customization_types = product_customizations.map { |product_customization| product_customization.product_customization_type }.uniq
+      product = variant.product
```

Comment:
Unless you ran into an issue where this was needed, `Spree::LineItem` already has an association to a `Spree::Product` through the `Spree::Variant`.